### PR TITLE
Fix bookmarks filter

### LINKED ISSUE #366: Filtering by tags throws InvalidFieldNameException
When selecting more than 1 tag the bookmarks list display gets corrupted as shown at the screenshot below:
bookmarks_issue

If multiple tags are selected all bookmarks are returned having any of the selected tags assigned instead of bookmarks having all selected tags assigned.

The issue seems to be with tag's name being surrounded with backticks characters in SQL query.

### Expected behaviour
- No exception logged in the log
- The tags assigned to the bookmark are visible only once
- Only bookmarks having all selected tags assigned are displayed

### Actual behaviour
Nextcloud shows the following log entry:
```
Doctrine\DBAL\Exception\InvalidFieldNameException: An exception occurred while executing 'SELECT t.`tag`, COUNT(t.`bookmark_id`) AS `nbr` FROM `oc_bookmarks_tags` t INNER JOIN `oc_bookmarks` b ON (b.`id` = t.`bookmark_id`) AND (b.`user_id` = ?) WHERE t.`tag` NOT IN (`dotNET`) GROUP BY t.`tag` ORDER BY `nbr` DESC LIMIT 18446744073709551615 OFFSET 0' with params ["pixi"]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'dotNET' in 'where clause'
```

### Server configuration

**Operating system**: Ubuntu 16.04 LTS

**Web server:** apache 2.4.18-2ubuntu3.2

**Database:** mysql  5.7.18-0ubuntu0.16.04.1

**PHP version:** 7.0.15-0ubuntu0.16.04.4

**Nextcloud version:** 11.0.3

**Bookmarks version:** 0.10.0

**Updated from an older Nextcloud or fresh install:**

**Signing status:**

```
No errors have been found.
```

**List of activated apps:**

```
Enabled:
  - activity: 2.4.1
  - bookmarks: 0.10.0
  - calendar: 1.5.3
  - comments: 1.1.0
  - contacts: 1.5.3
  - dav: 1.1.1
  - direct_menu: 0.10.2
  - external: true
  - federatedfilesharing: 1.1.1
  - federation: 1.1.1
  - files: 1.6.1
  - files_automatedtagging: 1.1.1
  - files_external: 1.1.2
  - files_markdown: 1.0.1
  - files_pdfviewer: 1.0.1
  - files_retention: 1.0.1
  - files_sharing: 1.1.1
  - files_texteditor: 2.2
  - files_trashbin: 1.1.0
  - files_versions: 1.4.0
  - files_videoplayer: 1.0.0
  - firstrunwizard: 2.0
  - gallery: 16.0.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.0
  - notes: 2.2.0
  - notifications: 1.0.1
  - ocsms: 1.12.0
  - password_policy: 1.1.0
  - previewgenerator: 1.0.6
  - provisioning_api: 1.1.0
  - qownnotesapi: 17.5.0
  - richdocuments: 1.11.27
  - serverinfo: 1.1.1
  - sharebymail: 1.0.1
  - survey_client: 0.1.5
  - systemtags: 1.1.3
  - tasks: 0.9.5
  - templateeditor: 0.2
  - theming: 1.1.1
  - twofactor_backupcodes: 1.0.0
  - updatenotification: 1.1.1
  - workflowengine: 1.1.1
Disabled:
  - admin_audit
  - encryption
  - files_accesscontrol
  - nextant
  - ownnote
  - user_external
  - user_ldap
  - user_saml
```

**Nextcloud configuration:**

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            ***REMOVED SENSITIVE VALUE***
        ],
        "datadirectory": "\/data\/nextcloud",
        "overwrite.cli.url": "http:\/\/***REMOVED SENSITIVE VALUE***\/nc",
        "dbtype": "mysql",
        "version": "11.0.3.2",
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "theme": "",
        "maintenance": false,
        "loglevel": 2,
        "enable_previews": true,
        "preview_libreoffice_path": "\/usr\/bin\/libreoffice",
        "trashbin_retention_obligation": "auto",
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "ssl",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "gmail.com",
        "mail_smtpauth": 1,
        "mail_smtphost": "smtp.gmail.com",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "cache_path": "\/var\/cache\/nextcloud",
        "mysql.utf8mb4": true
    }
}
```

**Are you using external storage, if yes which one:** smb

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Chrome 58.0.3029.110 (64-bit)

**Operating system:** Windows 10

#### Web server error log
```
[22/May/2017:21:50:19 +0200] "GET /nc/index.php/apps/bookmarks/bookmark?type=bookmark&tag=dotNET&page=0&sort=bookmarks_sorting_recent HTTP/1.1" 200 9231 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
[22/May/2017:21:50:19 +0200] "GET /nc/index.php/apps/bookmarks/bookmark?type=rel_tags&tag=dotNET&page=0&sort=bookmarks_sorting_recent HTTP/1.1" 500 22419 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
[22/May/2017:21:50:19 +0200] "GET /nc/index.php/apps/bookmarks/bookmark?type=bookmark&tag=dotNET&page=1&sort=bookmarks_sorting_recent HTTP/1.1" 200 4314 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
```

#### Nextcloud log (data/nextcloud.log)
```
`Doctrine\DBAL\Exception\InvalidFieldNameException: An exception occurred while executing 'SELECT t.`tag`, COUNT(t.`bookmark_id`) AS `nbr` FROM `oc_bookmarks_tags` t INNER JOIN `oc_bookmarks` b ON (b.`id` = t.`bookmark_id`) AND (b.`user_id` = ?) WHERE t.`tag` NOT IN (`dotNET`) GROUP BY t.`tag` ORDER BY `nbr` DESC LIMIT 18446744073709551615 OFFSET 0' with params ["pixi"]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'dotNET' in 'where clause'`

**Strack trace:**
`/data/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php - line 128: Doctrine\DBAL\Driver\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\DBAL\Driver\PDOException))
/data/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php - line 836: Doctrine\DBAL\DBALException driverExceptionDuringQuery(Object(Doctrine\DBAL\Driver\PDOMySql\Driver), Object(Doctrine\DBAL\Driver\PDOException), 'SELECT t.`tag`,...', Array)
/data/www/nextcloud/lib/private/DB/Connection.php - line 194: Doctrine\DBAL\Connection->executeQuery('SELECT t.`tag`,...', Array, Array, NULL)
/data/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php - line 206: OC\DB\Connection->executeQuery('SELECT t.`tag`,...', Array, Array)
/data/www/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php - line 177: Doctrine\DBAL\Query\QueryBuilder->execute()
/data/www/nextcloud/apps/bookmarks/controller/lib/bookmarks.php - line 94: OC\DB\QueryBuilder\QueryBuilder->execute()
/data/www/nextcloud/apps/bookmarks/controller/rest/bookmarkcontroller.php - line 97: OCA\Bookmarks\Controller\Lib\Bookmarks->findTags('pixi', Array)
/data/www/nextcloud/apps/bookmarks/controller/rest/internalbookmarkcontroller.php - line 56: OCA\Bookmarks\Controller\Rest\BookmarkController->getBookmarks('rel_tags', 'dotNET', 0, 'bookmarks_sorti...', 'pixi', Array, 'or', '', Array)
[internal function] OCA\Bookmarks\Controller\Rest\InternalBookmarkController->getBookmarks('rel_tags', 'dotNET', 0, 'bookmarks_sorti...', NULL, Array, 'or', '', Array)
/data/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php - line 160: call_user_func_array(Array, Array)
/data/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php - line 90: OC\AppFramework\Http\Dispatcher->executeController(Object(OCA\Bookmarks\Controller\Rest\InternalBookmarkController), 'getBookmarks')
/data/www/nextcloud/lib/private/AppFramework/App.php - line 114: OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\Bookmarks\Controller\Rest\InternalBookmarkController), 'getBookmarks')
/data/www/nextcloud/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47: OC\AppFramework\App main('InternalBookmar...', 'getBookmarks', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)
[internal function] OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)
/data/www/nextcloud/lib/private/Route/Router.php - line 299: call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)
/data/www/nextcloud/lib/base.php - line 1010: OC\Route\Router->match('/apps/bookmarks...')
/data/www/nextcloud/index.php - line 40: OC handleRequest()
{main}`
```

#### Browser log
```
```

### PR DESCRIPTION
**Pull request by marcelklehr**
used to use AND for different item columns as well which is plain wrong.
Also, client-side didn't supply a conjunction overriding the default OR (fixes #366).

### CODE REVIEW

**Comment by marcelklehr** [comment]
(nc master is failing here, too, due to nc v12 release, presumably)

**Comment by blizzz** [comment]
@pixi-p mind testing?

**Comment by marcelklehr** [comment]
I have a feeling this might fix #366 as well...

**Comment by marcelklehr** [comment]
Mmmh. @blizzz can you check out the latest travis build? Starting on nc12 my latest try to fix #366 is failing :/

**Comment by blizzz** [comment]
tests against master are running, and so far they look good https://travis-ci.org/nextcloud/bookmarks/builds/246725852

**Comment by marcelklehr** [comment]
Sorry. I meant, I don't understand why this pr fails on nc12+

**Comment by blizzz** [comment]
good question, it also runs here locally… i restarted tests :D going to investigate later again if it repeatedly fails.

**Comment by blizzz** [comment]
It only failes on sqlite. 

From debugging: the bookmark is added correctly, gets an ID.

So issue lies within the select. It looks like this:

```sql
SELECT "url",
       "title",
       GROUP_CONCAT("t"."tag") AS "tags"
FROM "oc_bookmarks" b
LEFT JOIN "oc_bookmarks_tags" t ON "t"."bookmark_id" = "b"."id"
WHERE "user_id" = :dcValue1
GROUP BY "url",
         "title",
         "lastmodified"
HAVING ("tags" LIKE :dcValue2 ESCAPE '\') OR ("tags" LIKE :dcValue3 ESCAPE '\')
ORDER BY "lastmodified" DESC
```

on succeeding MySQL it looks like this:

```sql
SELECT `url`,
       `title`,
       GROUP_CONCAT(`t`.`tag`) AS `tags`
FROM `oc_bookmarks` b
LEFT JOIN `oc_bookmarks_tags` t ON `t`.`bookmark_id` = `b`.`id`
WHERE `user_id` = :dcValue1
GROUP BY `url`,
         `title`,
         `lastmodified`
HAVING (`tags` LIKE :dcValue2)
OR (`tags` LIKE :dcValue3)
ORDER BY `lastmodified` DESC
```

The difference is the ``ESCAPE '\'`` in the SQLite operation which was introduced here https://github.com/nextcloud/server/commit/a65652fc1efe158bd097d573f012167cfcd26a62 (file lib/private/DB/QueryBuilder/ExpressionBuilder/SqliteExpressionBuilder.php) for Nc 12.

Removing this addition makes the tests pass.

But now, what is the issue with ``ESCAPE``? Especially since "one" and "three" are provided as arguments in this case, no escape character present. The doc does not state much more:

> If the optional ESCAPE clause is present, then the expression following the ESCAPE keyword must evaluate to a string consisting of a single character. This character may be used in the LIKE pattern to include literal percent or underscore characters. The escape character followed by a percent symbol (%), underscore (_), or a second instance of the escape character itself matches a literal percent symbol, underscore, or a single escape character, respectively. (https://sqlite.org/lang_expr.html#like)

For me it appears like unexpected and undesired behaviour from SQLite…

@icewind1991 any idea?

**Comment by blizzz** [comment]
@marcelklehr I found a workaround. Use ``iLike`` instead of ``like``.

**Comment by marcelklehr** [comment]
Nice! Postgres still complains, but due to a different problem. For some reason I'm unable to use the aggregated column in the having clause.

**Comment by marcelklehr** [comment]
:tada:

File: tests/lib_bookmark_test.php [hunk 1]
```diff
@@ -72,11 +72,12 @@ function testFindBookmarksSelectAndOrFilteredTags() {
 		$this->libBookmarks->addBookmark($secondUser, "http://www.heise.de", "Heise", array("one", "two"), "PrivatTag", false);
 		$this->libBookmarks->addBookmark($secondUser, "http://www.golem.de", "Golem", array("four"), "PublicNoTag", true);
 		$this->libBookmarks->addBookmark($secondUser, "http://www.9gag.com", "9gag", array("two", "three"), "PublicTag", true);
-		$resultSetOne = $this->libBookmarks->findBookmarks($this->userid, 0, 'lastmodified', array('one', 'three'), true, -1, false, array('url', 'title'), 'or');
-		$this->assertEquals(3, count($resultSetOne));
+		$resultSetOne = $this->libBookmarks->findBookmarks($this->userid, 0, 'lastmodified', array('one', 'three'), true, -1, false, array('url', 'title', 'tags'), 'or');
+		var_dump($resultSetOne);
```
Comment (blizzz) [inline tests/lib_bookmark_test.php]: can you remove the debug output? then we're good to merge 🎉

**Comment by blizzz** [comment]
good job @marcelklehr 🚀

**Pull request merged**